File: apps/api/internal/calculation/monthly.go
package calculation

import "github.com/shopspring/decimal"

// CreditType defines how overtime is credited to the flextime account (Art der Gutschrift).
type CreditType string

const (
	// CreditTypeNoEvaluation transfers overtime/undertime directly 1:1 with no limits.
	CreditTypeNoEvaluation CreditType = "no_evaluation"
	// CreditTypeCompleteCarryover transfers overtime with monthly and balance caps.
	CreditTypeCompleteCarryover CreditType = "complete_carryover"
	// CreditTypeAfterThreshold only credits overtime exceeding the threshold.
	CreditTypeAfterThreshold CreditType = "after_threshold"
	// CreditTypeNoCarryover resets the flextime balance to zero each month.
	CreditTypeNoCarryover CreditType = "no_carryover"
)

// MonthlyCalcInput contains all data needed for monthly aggregation calculation.
type MonthlyCalcInput struct {
	DailyValues       []DailyValueInput       // Daily values for the month
	PreviousCarryover int                     // Flextime balance from previous month (minutes)
	EvaluationRules   *MonthlyEvaluationInput // ZMI rules (nil = no evaluation)
	AbsenceSummary    AbsenceSummaryInput     // Pre-computed absence counts
}

// DailyValueInput represents a simplified daily value for monthly aggregation.
type DailyValueInput struct {
	Date       string // YYYY-MM-DD reference
	GrossTime  int    // Minutes
	NetTime    int    // Minutes
	TargetTime int    // Minutes
	Overtime   int    // Minutes (positive)
	Undertime  int    // Minutes (positive, to subtract)
	BreakTime  int    // Minutes
	HasError   bool
}

// MonthlyEvaluationInput contains ZMI monthly evaluation rules.
type MonthlyEvaluationInput struct {
	CreditType          CreditType // Which of the 4 credit types
	FlextimeThreshold   *int       // Threshold for after_threshold mode
	MaxFlextimePerMonth *int       // Monthly credit cap
	FlextimeCapPositive *int       // Upper balance limit
	FlextimeCapNegative *int       // Lower balance limit (stored as positive value)
	AnnualFloorBalance  *int       // Year-end annual floor
}

// AbsenceSummaryInput contains pre-computed absence data.
type AbsenceSummaryInput struct {
	VacationDays     decimal.Decimal
	SickDays         int
	OtherAbsenceDays int
}

// MonthlyCalcOutput contains the results of monthly aggregation calculation.
type MonthlyCalcOutput struct {
	// Aggregated totals (all in minutes)
	TotalGrossTime  int
	TotalNetTime    int
	TotalTargetTime int
	TotalOvertime   int
	TotalUndertime  int
	TotalBreakTime  int

	// Flextime tracking (all in minutes)
	FlextimeStart     int // PreviousCarryover
	FlextimeChange    int // TotalOvertime - TotalUndertime
	FlextimeRaw       int // FlextimeStart + FlextimeChange
	FlextimeCredited  int // Amount actually credited after rules
	FlextimeForfeited int // Amount forfeited due to rules
	FlextimeEnd       int // Final balance after all rules

	// Work summary
	WorkDays       int
	DaysWithErrors int

	// Absence copy
	VacationTaken    decimal.Decimal
	SickDays         int
	OtherAbsenceDays int

	// Warnings
	Warnings []string
}

// CalculateMonth aggregates daily values into monthly totals and applies
// ZMI-compliant credit type rules for flextime carryover.
func CalculateMonth(input MonthlyCalcInput) MonthlyCalcOutput {
	var output MonthlyCalcOutput

	// Step 1: Initialize
	output.FlextimeStart = input.PreviousCarryover
	output.VacationTaken = input.AbsenceSummary.VacationDays
	output.SickDays = input.AbsenceSummary.SickDays
	output.OtherAbsenceDays = input.AbsenceSummary.OtherAbsenceDays
	output.Warnings = make([]string, 0)

	// Step 2: Aggregate daily values
	for _, dv := range input.DailyValues {
		output.TotalGrossTime += dv.GrossTime
		output.TotalNetTime += dv.NetTime
		output.TotalTargetTime += dv.TargetTime
		output.TotalOvertime += dv.Overtime
		output.TotalUndertime += dv.Undertime
		output.TotalBreakTime += dv.BreakTime

		if dv.GrossTime > 0 || dv.NetTime > 0 {
			output.WorkDays++
		}
		if dv.HasError {
			output.DaysWithErrors++
		}
	}

	// Step 3: Calculate flextime change
	output.FlextimeChange = output.TotalOvertime - output.TotalUndertime

	// Step 4: Calculate raw flextime
	output.FlextimeRaw = output.FlextimeStart + output.FlextimeChange

	// Step 5: Apply credit type rules
	if input.EvaluationRules != nil {
		output = applyCreditType(output, *input.EvaluationRules)
	} else {
		// No evaluation: direct transfer
		output.FlextimeCredited = output.FlextimeChange
		output.FlextimeEnd = output.FlextimeRaw
		output.FlextimeForfeited = 0
	}

	return output
}

// applyCreditType implements the 4 ZMI credit types for flextime calculation.
func applyCreditType(output MonthlyCalcOutput, rules MonthlyEvaluationInput) MonthlyCalcOutput {
	switch rules.CreditType {
	case CreditTypeNoEvaluation:
		output.FlextimeCredited = output.FlextimeChange
		output.FlextimeEnd = output.FlextimeRaw
		output.FlextimeForfeited = 0

	case CreditTypeCompleteCarryover:
		credited := output.FlextimeChange

		// Apply monthly cap
		if rules.MaxFlextimePerMonth != nil && credited > *rules.MaxFlextimePerMonth {
			output.FlextimeForfeited = credited - *rules.MaxFlextimePerMonth
			credited = *rules.MaxFlextimePerMonth
			output.Warnings = append(output.Warnings, WarnCodeMonthlyCap)
		}

		output.FlextimeCredited = credited
		output.FlextimeEnd = output.FlextimeStart + credited

		// Apply positive/negative caps
		prevEnd := output.FlextimeEnd
		var forfeited int
		output.FlextimeEnd, forfeited = applyFlextimeCaps(output.FlextimeEnd, rules.FlextimeCapPositive, rules.FlextimeCapNegative)
		output.FlextimeForfeited += forfeited
		if output.FlextimeEnd != prevEnd {
			output.Warnings = append(output.Warnings, WarnCodeFlextimeCapped)
		}

	case CreditTypeAfterThreshold:
		threshold := 0
		if rules.FlextimeThreshold != nil {
			threshold = *rules.FlextimeThreshold
		}

		if output.FlextimeChange > threshold {
			// Above threshold: credit the excess
			output.FlextimeCredited = output.FlextimeChange - threshold
			output.FlextimeForfeited = threshold
		} else if output.FlextimeChange > 0 {
			// Positive but at or below threshold: forfeit all
			output.FlextimeCredited = 0
			output.FlextimeForfeited = output.FlextimeChange
			output.Warnings = append(output.Warnings, WarnCodeBelowThreshold)
		} else {
			// Undertime: fully deduct (no threshold applies to undertime)
			output.FlextimeCredited = output.FlextimeChange
			output.FlextimeForfeited = 0
		}

		// Apply monthly cap
		if rules.MaxFlextimePerMonth != nil && output.FlextimeCredited > *rules.MaxFlextimePerMonth {
			excess := output.FlextimeCredited - *rules.MaxFlextimePerMonth
			output.FlextimeForfeited += excess
			output.FlextimeCredited = *rules.MaxFlextimePerMonth
			output.Warnings = append(output.Warnings, WarnCodeMonthlyCap)
		}

		output.FlextimeEnd = output.FlextimeStart + output.FlextimeCredited

		// Apply positive/negative caps
		prevEnd := output.FlextimeEnd
		var forfeited int
		output.FlextimeEnd, forfeited = applyFlextimeCaps(output.FlextimeEnd, rules.FlextimeCapPositive, rules.FlextimeCapNegative)
		output.FlextimeForfeited += forfeited
		if output.FlextimeEnd != prevEnd {
			output.Warnings = append(output.Warnings, WarnCodeFlextimeCapped)
		}

	case CreditTypeNoCarryover:
		output.FlextimeCredited = 0
		output.FlextimeEnd = 0
		output.FlextimeForfeited = output.FlextimeChange
		output.Warnings = append(output.Warnings, WarnCodeNoCarryover)

	default:
		// Unknown credit type: default to no evaluation (direct transfer)
		output.FlextimeCredited = output.FlextimeChange
		output.FlextimeEnd = output.FlextimeRaw
		output.FlextimeForfeited = 0
	}

	return output
}

// applyFlextimeCaps applies positive and negative balance caps.
// Returns the capped value and additional forfeited amount.
func applyFlextimeCaps(flextime int, capPositive, capNegative *int) (int, int) {
	forfeited := 0

	if capPositive != nil && flextime > *capPositive {
		forfeited = flextime - *capPositive
		flextime = *capPositive
	}

	if capNegative != nil && flextime < -*capNegative {
		flextime = -*capNegative
	}

	return flextime, forfeited
}

// CalculateAnnualCarryover determines the year-end carryover with annual floor.
// If currentBalance is nil, returns 0. If annualFloor is set and the balance is
// below the negative floor, the floor is applied.
func CalculateAnnualCarryover(currentBalance, annualFloor *int) int {
	if currentBalance == nil {
		return 0
	}
	balance := *currentBalance
	if annualFloor != nil && balance < -*annualFloor {
		return -*annualFloor
	}
	return balance
}

File: src/lib/calculation/monthly.ts
/**
 * Monthly Calculation Engine
 *
 * Pure-function module for monthly aggregation and flextime credit type rules.
 * No database access, no side effects.
 *
 * Ported from Go: apps/api/internal/calculation/monthly.go (250 lines)
 */

import type { Decimal } from "@prisma/client/runtime/client"
import {
  WARN_MONTHLY_CAP_REACHED,
  WARN_FLEXTIME_CAPPED,
  WARN_BELOW_THRESHOLD,
  WARN_NO_CARRYOVER,
} from "./errors"

// --- Credit Type ---

/** CreditType defines how overtime is credited to the flextime account (Art der Gutschrift). */
export type CreditType =
  | "no_evaluation"
  | "complete_carryover"
  | "after_threshold"
  | "no_carryover"

/** Transfers overtime/undertime directly 1:1 with no limits. */
export const CREDIT_TYPE_NO_EVALUATION: CreditType = "no_evaluation"
/** Transfers overtime with monthly and balance caps. */
export const CREDIT_TYPE_COMPLETE_CARRYOVER: CreditType = "complete_carryover"
/** Only credits overtime exceeding the threshold. */
export const CREDIT_TYPE_AFTER_THRESHOLD: CreditType = "after_threshold"
/** Resets the flextime balance to zero each month. */
export const CREDIT_TYPE_NO_CARRYOVER: CreditType = "no_carryover"

// --- Input / Output Types ---

/** Simplified daily value for monthly aggregation. */
export interface DailyValueInput {
  date: string // YYYY-MM-DD
  grossTime: number // Minutes
  netTime: number // Minutes
  targetTime: number // Minutes
  overtime: number // Minutes (positive)
  undertime: number // Minutes (positive, to subtract)
  breakTime: number // Minutes
  hasError: boolean
}

/** ZMI monthly evaluation rules. */
export interface MonthlyEvaluationInput {
  creditType: CreditType
  flextimeThreshold: number | null // Threshold for after_threshold mode
  maxFlextimePerMonth: number | null // Monthly credit cap
  flextimeCapPositive: number | null // Upper balance limit
  flextimeCapNegative: number | null // Lower balance limit (stored as positive value)
  annualFloorBalance: number | null // Year-end annual floor
}

/** Pre-computed absence data. */
export interface AbsenceSummaryInput {
  vacationDays: Decimal
  sickDays: number
  otherAbsenceDays: number
}

/** All data needed for monthly aggregation calculation. */
export interface MonthlyCalcInput {
  dailyValues: DailyValueInput[]
  previousCarryover: number // Flextime balance from previous month (minutes)
  evaluationRules: MonthlyEvaluationInput | null // null = no evaluation
  absenceSummary: AbsenceSummaryInput
}

/** Results of monthly aggregation calculation. */
export interface MonthlyCalcOutput {
  // Aggregated totals (all in minutes)
  totalGrossTime: number
  totalNetTime: number
  totalTargetTime: number
  totalOvertime: number
  totalUndertime: number
  totalBreakTime: number

  // Flextime tracking (all in minutes)
  flextimeStart: number // PreviousCarryover
  flextimeChange: number // TotalOvertime - TotalUndertime
  flextimeRaw: number // FlextimeStart + FlextimeChange
  flextimeCredited: number // Amount actually credited after rules
  flextimeForfeited: number // Amount forfeited due to rules
  flextimeEnd: number // Final balance after all rules

  // Work summary
  workDays: number
  daysWithErrors: number

  // Absence copy
  vacationTaken: Decimal
  sickDays: number
  otherAbsenceDays: number

  // Warnings
  warnings: string[]
}

// --- Public Functions ---

/**
 * Aggregates daily values into monthly totals and applies
 * ZMI-compliant credit type rules for flextime carryover.
 */
export function calculateMonth(input: MonthlyCalcInput): MonthlyCalcOutput {
  const output: MonthlyCalcOutput = {
    totalGrossTime: 0,
    totalNetTime: 0,
    totalTargetTime: 0,
    totalOvertime: 0,
    totalUndertime: 0,
    totalBreakTime: 0,
    flextimeStart: input.previousCarryover,
    flextimeChange: 0,
    flextimeRaw: 0,
    flextimeCredited: 0,
    flextimeForfeited: 0,
    flextimeEnd: 0,
    workDays: 0,
    daysWithErrors: 0,
    vacationTaken: input.absenceSummary.vacationDays,
    sickDays: input.absenceSummary.sickDays,
    otherAbsenceDays: input.absenceSummary.otherAbsenceDays,
    warnings: [],
  }

  // Step 2: Aggregate daily values
  for (const dv of input.dailyValues) {
    output.totalGrossTime += dv.grossTime
    output.totalNetTime += dv.netTime
    output.totalTargetTime += dv.targetTime
    output.totalOvertime += dv.overtime
    output.totalUndertime += dv.undertime
    output.totalBreakTime += dv.breakTime

    if (dv.grossTime > 0 || dv.netTime > 0) {
      output.workDays++
    }
    if (dv.hasError) {
      output.daysWithErrors++
    }
  }

  // Step 3: Calculate flextime change
  output.flextimeChange = output.totalOvertime - output.totalUndertime

  // Step 4: Calculate raw flextime
  output.flextimeRaw = output.flextimeStart + output.flextimeChange

  // Step 5: Apply credit type rules
  if (input.evaluationRules !== null) {
    return applyCreditType(output, input.evaluationRules)
  } else {
    // No evaluation: direct transfer
    output.flextimeCredited = output.flextimeChange
    output.flextimeEnd = output.flextimeRaw
    output.flextimeForfeited = 0
  }

  return output
}

/**
 * Implements the 4 ZMI credit types for flextime calculation.
 */
function applyCreditType(
  output: MonthlyCalcOutput,
  rules: MonthlyEvaluationInput,
): MonthlyCalcOutput {
  switch (rules.creditType) {
    case "no_evaluation": {
      output.flextimeCredited = output.flextimeChange
      output.flextimeEnd = output.flextimeRaw
      output.flextimeForfeited = 0
      break
    }

    case "complete_carryover": {
      let credited = output.flextimeChange

      // Apply monthly cap
      if (
        rules.maxFlextimePerMonth !== null &&
        credited > rules.maxFlextimePerMonth
      ) {
        output.flextimeForfeited = credited - rules.maxFlextimePerMonth
        credited = rules.maxFlextimePerMonth
        output.warnings.push(WARN_MONTHLY_CAP_REACHED)
      }

      output.flextimeCredited = credited
      output.flextimeEnd = output.flextimeStart + credited

      // Apply positive/negative caps
      const prevEnd = output.flextimeEnd
      const capResult = applyFlextimeCaps(
        output.flextimeEnd,
        rules.flextimeCapPositive,
        rules.flextimeCapNegative,
      )
      output.flextimeEnd = capResult.value
      output.flextimeForfeited += capResult.forfeited
      if (output.flextimeEnd !== prevEnd) {
        output.warnings.push(WARN_FLEXTIME_CAPPED)
      }
      break
    }

    case "after_threshold": {
      const threshold =
        rules.flextimeThreshold !== null ? rules.flextimeThreshold : 0

      if (output.flextimeChange > threshold) {
        // Above threshold: credit the excess
        output.flextimeCredited = output.flextimeChange - threshold
        output.flextimeForfeited = threshold
      } else if (output.flextimeChange > 0) {
        // Positive but at or below threshold: forfeit all
        output.flextimeCredited = 0
        output.flextimeForfeited = output.flextimeChange
        output.warnings.push(WARN_BELOW_THRESHOLD)
      } else {
        // Undertime: fully deduct (no threshold applies to undertime)
        output.flextimeCredited = output.flextimeChange
        output.flextimeForfeited = 0
      }

      // Apply monthly cap
      if (
        rules.maxFlextimePerMonth !== null &&
        output.flextimeCredited > rules.maxFlextimePerMonth
      ) {
        const excess = output.flextimeCredited - rules.maxFlextimePerMonth
        output.flextimeForfeited += excess
        output.flextimeCredited = rules.maxFlextimePerMonth
        output.warnings.push(WARN_MONTHLY_CAP_REACHED)
      }

      output.flextimeEnd = output.flextimeStart + output.flextimeCredited

      // Apply positive/negative caps
      const prevEnd = output.flextimeEnd
      const capResult = applyFlextimeCaps(
        output.flextimeEnd,
        rules.flextimeCapPositive,
        rules.flextimeCapNegative,
      )
      output.flextimeEnd = capResult.value
      output.flextimeForfeited += capResult.forfeited
      if (output.flextimeEnd !== prevEnd) {
        output.warnings.push(WARN_FLEXTIME_CAPPED)
      }
      break
    }

    case "no_carryover": {
      output.flextimeCredited = 0
      output.flextimeEnd = 0
      output.flextimeForfeited = output.flextimeChange
      output.warnings.push(WARN_NO_CARRYOVER)
      break
    }

    default: {
      // Unknown credit type: default to no evaluation (direct transfer)
      output.flextimeCredited = output.flextimeChange
      output.flextimeEnd = output.flextimeRaw
      output.flextimeForfeited = 0
      break
    }
  }

  return output
}

/**
 * Applies positive and negative balance caps.
 * Returns the capped value and additional forfeited amount.
 */
export function applyFlextimeCaps(
  flextime: number,
  capPositive: number | null,
  capNegative: number | null,
): { value: number; forfeited: number } {
  let value = flextime
  let forfeited = 0

  if (capPositive !== null && value > capPositive) {
    forfeited = value - capPositive
    value = capPositive
  }

  if (capNegative !== null && value < -capNegative) {
    value = -capNegative
  }

  return { value, forfeited }
}

/**
 * Determines the year-end carryover with annual floor.
 * If currentBalance is null, returns 0. If annualFloor is set and the balance
 * is below the negative floor, the floor is applied.
 */
export function calculateAnnualCarryover(
  currentBalance: number | null,
  annualFloor: number | null,
): number {
  if (currentBalance === null) {
    return 0
  }
  const balance = currentBalance
  if (annualFloor !== null && balance < -annualFloor) {
    return -annualFloor
  }
  return balance
}
